File: leetcode_problems/python/lc106_binary_tree_from_postorder_and_inorder.py
from utils import TreeNode


class Solution:
    def buildTree(self, inorder: list[int], postorder: list[int]) -> TreeNode | None:
        def build(inorder: list[int], postorder: list[int]) -> TreeNode | None:
            if len(inorder) == 0 or len(postorder) == 0:
                return None

            rootVal = postorder.pop()
            root = TreeNode(rootVal)
            inPos = inorder.index(rootVal)

            root.right = build(inorder[inPos + 1 :], postorder)
            root.left = build(inorder[:inPos], postorder)
            return root

        # Оптимизация с маппером не работает, т.к. мы режем массив
        # На каждом шаге рекурсии у нас меняются индексы массива
        return build(inorder, postorder)

File: leetcode_problems/go/lc106_binary_tree_from_postorder_and_inorder.go
package main

func buildTreePostorderInorderHelper(postorder []int, postStart, postEnd int, inorder []int, inStart, inEnd int, inMap map[int]int) *TreeNode {
	if postStart < postEnd || inStart > inEnd {
		return nil
	}

	rootVal := postorder[postStart]
	root := &TreeNode{Val: rootVal}
	inRoot := inMap[rootVal]
	rightSize := inEnd - inRoot

	root.Right = buildTreePostorderInorderHelper(postorder, postStart-1, postStart-rightSize, inorder, inRoot+1, inEnd, inMap)
	root.Left = buildTreePostorderInorderHelper(postorder, postStart-rightSize-1, postEnd, inorder, inStart, inRoot-1, inMap)
	return root
}

func buildTreePostorderInorder(inorder []int, postorder []int) *TreeNode {
	inMap := make(map[int]int)
	for i, v := range inorder {
		inMap[v] = i
	}
	return buildTreePostorderInorderHelper(postorder, len(postorder)-1, 0, inorder, 0, len(inorder)-1, inMap)
}
